src/app/modules/submission/submission.utils.ts

```diff
@@ -68,3 +92,237 @@ export const extractEmailConfirmationData = (
     return acc
   }, [])
 }
+
+/**
+ * Extracts response data to be sent in email confirmations
+ * @param responses Responses from form filler
+ * @param formFields Fields from form object
+ * @returns Array of data for email confirmations
+ */
+export const extractEmailConfirmationDataFromIncomingSubmission = (
+  incomingSubmission: IncomingSubmission,
+): AutoReplyMailData[] => {
+  const { responses, form } = incomingSubmission
+  return extractEmailConfirmationData(responses, form.form_fields)
+}
+
+/**
+ * Filter allowed form field responses from given responses and return the
+ * array of responses with duplicates removed.
+ *
+ * @param form The form document
+ * @param responses the responses that corresponds to the given form
+ * @returns neverthrow ok() filtered list of allowed responses with duplicates (if any) removed
+ * @returns neverthrow err(ConflictError) if the given form's form field ids count do not match given responses'
+ */
+export const getFilteredResponses = (
+  form: IFormDocument,
+  responses: FieldResponse[],
+): Result<FieldResponse[], ConflictError> => {
+  const modeFilter = getModeFilter(form.responseMode)
+
+  if (!form.form_fields) {
+    return err(new ConflictError('Form fields are missing'))
+  }
+  // _id must be transformed to string as form response is jsonified.
+  const fieldIds = modeFilter(form.form_fields).map((field) => ({
+    _id: String(field._id),
+  }))
+  const uniqueResponses = _.uniqBy(modeFilter(responses), '_id')
+  const results = _.intersectionBy(uniqueResponses, fieldIds, '_id')
+
+  if (results.length < fieldIds.length) {
+    const onlyInForm = _.differenceBy(fieldIds, results, '_id').map(
+      ({ _id }) => _id,
+    )
+    return err(
+      new ConflictError('Some form fields are missing', {
+        formId: form._id,
+        onlyInForm,
+      }),
+    )
+  }
+  return ok(results)
+}
+
+export class IncomingSubmission {
+  constructor(
```

Comment:
privatise if not used?

Reply:
cant make private cus it's needed for child class construction. but marking it protected works, so i'll do that